Unique and Exist validator fixes

samdark opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #13842, #14042

vladis84 committed:
Fixed ambiguous error sql while using unique validator

samdark committed:
Refactored solution to be contained within ActiveQuery instead of helper

samdark committed:
Removed unneeded changelog line [skip ci]

samdark commented on framework/db/ActiveQuery.php:
```diff
@@ -796,4 +789,89 @@ public function alias($alias)
+     * @throws \yii\base\InvalidConfigException
+     * @since 2.0.12
+     */
+    public function getFromTableNames()
```
It's not used for a fix. Not sure it needs to be added at all.

samdark commented on framework/db/ActiveQuery.php:
```diff
@@ -796,4 +789,89 @@ public function alias($alias)
+     * @throws \yii\base\InvalidConfigException
+     * @since 2.0.12
+     */
+    public function getFromAliases()
```
Only first alias is actually used in unique validator. Not sure we need all aliases at all.

klimov-paul replied:
Method name does not sound good: `from` word sounds like preposition instead of noun here.

samdark replied:
Any suggestions for a better name?

SilverFire replied:
Agree with @klimov-paul - it's a bad name, especially for the public API.

samdark commented on framework/db/ActiveQuery.php:
```diff
@@ -796,4 +789,89 @@ public function alias($alias)
+     * @return string primary table name
+     * @since 2.0.12
+     */
+    protected function getPrimaryTableName()
```
This one is done for refactoring only. Could be left as it was before.

klimov-paul replied:
I would rather remove it

Faryshta replied:
i think staying DRY is better and keeping it.

samdark committed:
Removed obsolete import

samdark commented:
@vladis84 check the code. That's what I was talking about.

klimov-paul commented:
Same fix should be added to `ExistValidator` as well, as it has the same problem.

cebe commented:
This implements parts of what is discussed in #7263

samdark committed:
Applied disambiguation to ExistValidator as well

samdark committed:
Fixed changelog [skip ci]

samdark commented:
@klimov-paul added fix for `ExistValidator`.

SilverFire commented on framework/db/ActiveQuery.php:
```diff
@@ -796,4 +789,89 @@ public function alias($alias)
+        }
+
+        if (is_string($this->from)) {
+           $tableNames = preg_split('/\s*,\s*/', trim($this->from), -1, PREG_SPLIT_NO_EMPTY);
```
This should not happen in normal conditions. Method `from()` normalizes string to array before butting it to `$this->from`

SilverFire committed:
PHPDocs and code style fixed

rob006 commented:
This could be simplified a lot if #11646 will be merged.

samdark commented:
@cebe can you take a look at #11646?

cebe commented:
Sure, but its not trivial to handle. I already spent a lot of time on that issue and finding a good solution is quite hard.

samdark commented:
Then I'm going to merge this one as is.

samdark committed:
Alternative names

samdark commented:
Adjusted method names.

cebe commented:
no need to hurry with this.

dynasource commented on tests/framework/db/ActiveQueryTest.php:
```diff
@@ -231,4 +231,82 @@ public function testAlias_yet_set()
         $this->assertInstanceOf('yii\db\ActiveQuery', $result);
         $this->assertEquals(['alias' => 'old'], $result->from);
     }
+
```
dataProvider seems logical here?

samdark replied:
Yes and no. These are alike cases but checks vary so I'd keep it as is.

dynasource commented on framework/db/ActiveQuery.php:
```diff
@@ -796,4 +789,89 @@ public function alias($alias)
+     * @since 2.0.12
+     */
+    public function getTableNamesUsedInFrom()
+    {
```
this method is the only one about getting tableNames. Shouldnt this get a cleaner name like 'getTableNames' or 'extractTableNames'

samdark replied:
There are table names used in sub-queries and other cases. The method works only on `from` clause.

dynasource commented on framework/db/ActiveQuery.php:
```diff
@@ -796,4 +789,89 @@ public function alias($alias)
+     * @since 2.0.12
+     */
+    public function getAliasesUsedInFrom()
+    {
```
same here. The question is how durable such naming is for publishing it in the public api. Does it matter if it is called 'getAliases' to show 'what is does' instead of 'how it does it'?

samdark replied:
`usedInFrom` is important part of what it does. It returns aliases for `from` clause and not for `join` or subqueries.

cebe replied:
what is the purpose of this method? it returns alias names but does not relate them to the table names as far as I see. Wouldn't it be better to have one method that returns an array of 'alias' => 'tablename' entries. Just having a list of aliases without reference seems not useful to me.

Faryshta commented on framework/validators/ExistValidator.php:
```diff
@@ -137,14 +139,41 @@ private function prepareConditions($targetAttribute, $model, $attribute)
+
+        /** @var ActiveRecord $targetClass */
+        $query = $targetClass::find();
+        $tableAliases = $query->getAliasesUsedInFrom();
```
`getAliasesUsedInFrom()` is not part of `ActiveQueryInterface`. should it be added there?

samdark replied:
@Faryshta it belongs to SQL DBs only so I think it's OK.

Faryshta replied:
ok then. so you mean ExistValidator doesn't work with mongo?

samdark replied:
It will work with mongo well. See line 150:

```php
if (!$model instanceof ActiveRecord) {
    return $conditions;
}
```

ActiveRecord in this code is `\yii\db\ActiveRecord`.

cebe reviewed (changes requested):
This one can be done without affecting the solution to #7263 much. But I'd like to evaluate the implementation using one method instead of two, as pointed out in my comment: https://github.com/yiisoft/yii2/pull/14049#discussion_r115429569

samdark committed:
Further changes for unique and exist validators

- Combined methods for getting names and aliases of from tables
- Normalized names and aliases
- Added MSSQL syntax
- Added support for spaces in aliases and table names

samdark commented:
@cebe done.

cebe committed:
Fixed ambiguous column name in SELECT in UniqueValidator

fixes #14042

samdark committed:
Added escaping for column name

dynasource commented on framework/validators/ExistValidator.php:
```diff
@@ -137,14 +139,41 @@ private function prepareConditions($targetAttribute, $model, $attribute)
+            return $conditions;
+        }
+
+        // Add table prefix for column
```
this comment looks to be bound solely for the next line and is confusing. A private method for the whole next block would improve readability

dynasource commented on framework/validators/UniqueValidator.php:
```diff
@@ -249,14 +254,18 @@ private function prepareConditions($targetAttribute, $model, $attribute)
         foreach ($conditions as $columnName => $columnValue) {
-            $prefixedColumnName = "{$tableName}.$columnName";
-            $conditionsWithTableName[$prefixedColumnName] = $columnValue;
+            $prefixedColumn = "{{{$primaryTableAlias}}}.[[{$columnName}]]";
```
this is repeated logic, implement DRY?

dynasource commented on framework/validators/ExistValidator.php:
```diff
@@ -137,14 +139,41 @@ private function prepareConditions($targetAttribute, $model, $attribute)
+        }
+
+        // Add table prefix for column
+        $targetClass = $this->getTargetClass($model);
```
$targetModelClass would improve readability

dynasource reviewed (changes requested):
last critical review

samdark committed:
Additional refactoring as suggested by @dynasource

samdark commented:
@dynasource done.

samdark committed:
Merge branch 'master' into unique-validator-fix

samdark committed:
Fixed changelog [skip ci]
